src/platform/windows/misc.cpp

```diff
@@ -200,43 +202,78 @@ namespace platf {
+
+    // Get the session ID of the active console session
+    consoleSessionId = WTSGetActiveConsoleSessionId();
+    if (0xFFFFFFFF == consoleSessionId) {
+      // If there is no active console session, log a warning and return null
+      BOOST_LOG(warning) << "There isn't an active user session, therefore it is not possible to execute commands under the users profile.";
+      return nullptr;
     }
 
-    // Open a handle to the explorer.exe process
-    DWORD shell_pid;
-    GetWindowThreadProcessId(shell_window, &shell_pid);
-    HANDLE shell_process = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, FALSE, shell_pid);
-    if (!shell_process) {
-      BOOST_LOG(error) << "Failed to open shell process: "sv << GetLastError();
-      return NULL;
+    // Get the user token for the active console session
+    if (!WTSQueryUserToken(consoleSessionId, &userToken)) {
+      BOOST_LOG(debug) << "QueryUserToken failed, this would prevent commands from launching under the users profile.";
+      CloseHandle(userToken);
+      return nullptr;
     }
 
-    // Open explorer's token to clone for process creation
-    HANDLE shell_token;
-    BOOL ret = OpenProcessToken(shell_process, TOKEN_DUPLICATE, &shell_token);
-    CloseHandle(shell_process);
-    if (!ret) {
-      BOOST_LOG(error) << "Failed to open shell process token: "sv << GetLastError();
-      return NULL;
+    // Close the userToken handle when it goes out of scope
+    auto token_close = util::fail_guard([&userToken]() {
+      CloseHandle(userToken);
+    });
+
+    // We need to know if this is an elevated token or not.
+    // Get the elevation type of the user token
+    if (!GetTokenInformation(userToken, TokenElevationType, &elevationType, sizeof(TOKEN_ELEVATION_TYPE), &dwSize)) {
+      BOOST_LOG(debug) << "Retrieving token information failed: " << GetLastError();
+      CloseHandle(userToken);
+      return nullptr;
     }
 
-    // Duplicate the token to make it usable for process creation
-    HANDLE new_token;
-    ret = DuplicateTokenEx(shell_token, TOKEN_ALL_ACCESS, NULL, SecurityImpersonation, TokenPrimary, &new_token);
-    CloseHandle(shell_token);
-    if (!ret) {
-      BOOST_LOG(error) << "Failed to duplicate shell process token: "sv << GetLastError();
-      return NULL;
+    // User has a limited token, this likely means they have UAC enabled.
+    if (elevated && elevationType == TokenElevationTypeLimited) {
+      TOKEN_LINKED_TOKEN linkedToken;
+      // Retrieve the administrator token that is linked to the limited token
+      if (!GetTokenInformation(userToken, TokenLinkedToken, reinterpret_cast<void *>(&linkedToken), sizeof(TOKEN_LINKED_TOKEN), &dwSize)) {
+        // If the retrieval failed, log an error message and return null
+        BOOST_LOG(debug) << "Retrieving linked token information failed: " << GetLastError();
+
+        //Maybe the user is not an admin here, so let's just return a normal token.
+        CloseHandle(linkedToken.LinkedToken);
```

Comment:
Same here, we shouldn't close on failure.

The general contract of most Windows APIs is that any output parameters are in an undefined state on failure, so we shouldn't be touching them. Internally, if the token was somehow received prior to another error, it will take care of closing the handle for you.